dcos/nodeutil/util.go

```diff
@@ -0,0 +1,325 @@
+		return nil, err
+	}
+
+	ce, err := exec.Run("sh", []string{d.detectIPLocation}, exec.Timeout(time.Second*5))
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := ioutil.ReadAll(ce)
+	if err != nil {
+		return nil, err
+	}
+
+	err = <-ce.Done
+	if err != nil {
+		return nil, err
+	}
+
+	// strip the trailing \n
+	detectedIP := string(bytes.TrimSpace(buf))
+	if detectedIP == "" {
+		return nil, ErrNodeUtil{fmt.Sprintf("command %s return empty output", dcos.FileDetectIP)}
+	}
+
+	validIP := net.ParseIP(detectedIP)
+	if validIP == nil {
+		return nil, ErrNodeUtil{fmt.Sprintf("command %s returned invalid IP address %s", dcos.FileDetectIP, detectedIP)}
+	}
+
+	// save retrieved IP address to cache.
+	if d.cache {
+		d.Lock()
+		d.cachedIP = &validIP
+		d.Unlock()
+	}
+
+	return validIP, nil
+}
+
+// Role return a node role.
+func (d *dcosInfo) Role() (string, error) {
+	// return cached node role
+	if d.cache && d.cachedRole != "" {
+		return d.cachedRole, nil
+	}
+
+	isFile := func(path string) bool {
+		if _, err := os.Stat(path); err == nil {
+			return true
+		}
+		return false
+	}
+
+	// iterate over files and get the role.
+	var roles []string
+	for _, i := range []struct {
+		file string
+		role string
+	}{
+		{
+			file: d.roleMasterFile,
+			role: dcos.RoleMaster,
+		},
+		{
+			file: d.roleAgentFile,
+			role: dcos.RoleAgent,
+		},
+		{
+			file: d.roleAgentPublicFile,
+			role: dcos.RoleAgentPublic,
+		},
+	} {
+		if ok := isFile(i.file); ok {
+			roles = append(roles, i.role)
+		}
+	}
+
+	if len(roles) == 0 || len(roles) > 1 {
+		return "", ErrNodeUtil{fmt.Sprintf("Node must have only one role. Got %s", roles)}
+	}
+
+	if d.cache {
+		d.Lock()
+		d.cachedRole = roles[0]
+		d.Unlock()
+	}
+
+	return roles[0], nil
+}
+
+// IsLeader checks if the node is leader.
+func (d *dcosInfo) IsLeader() (bool, error) {
+	if d.cache && d.cachedIsLeader != nil {
+		return *d.cachedIsLeader, nil
+	}
+
+	// find role and IP before locking the structure.
+	role, err := d.Role()
+	if err != nil {
+		return false, err
+	}
+
+	localIP, err := d.DetectIP()
+	if err != nil {
+		return false, err
+	}
+
+	// agent cannot be leader
+	if role != dcos.RoleMaster {
+		return false, nil
+	}
+
+	addrs, err := net.LookupIP(d.dnsRecordLeader)
+	if err != nil {
+		return false, err
+	}
+
+	for _, addr := range addrs {
+		if localIP.Equal(addr) {
+			isLeader := true
+
+			if d.cache {
+				d.Lock()
+				d.cachedIsLeader = &isLeader
+				d.Unlock()
+			}
+
+			return isLeader, nil
+		}
+	}
+
+	return false, ErrNodeUtil{fmt.Sprintf("Error getting mesos leader. Number of ip addresses %d", len(addrs))}
+}
+
+// MesosID returns a mesosID for leading master and agents.
+func (d *dcosInfo) MesosID() (string, error) {
+	if d.client == nil {
+		panic("Unable to get mesos ID. Uninitialized http client")
+	}
+
+	if d.mesosStateURL == "" {
+		panic("Unable to get mesos ID. Uninitialized url")
+	}
+
+	req, err := http.NewRequest("GET", d.mesosStateURL, nil)
```

Comment:
are you ever worried about timeouts here? or do you expect that the caller has already configured the http client as needed? might be useful to accept a Context parameter for this func